kvbc/src/categorization/block_merkle_category.cpp

```diff
@@ -0,0 +1,398 @@
+        children[i] = sparse_merkle::LeafChild(sm_hash, sm_key);
+      } else {
+        auto internal_child = std::get<InternalChild>(child);
+        auto sm_hash = sparse_merkle::Hash(internal_child.hash);
+        children[i] = sparse_merkle::InternalChild{sm_hash, internal_child.version};
+      }
+    }
+  }
+  return sparse_merkle::BatchedInternalNode(children);
+}
+
+std::vector<Hash> hashedKeys(const std::vector<std::string>& keys) {
+  std::vector<Hash> hashed_keys;
+  hashed_keys.reserve(keys.size());
+  std::transform(keys.begin(), keys.end(), std::back_inserter(hashed_keys), [](auto& key) {
+    return Hasher{}.digest(key.data(), key.size());
+  });
+  return hashed_keys;
+}
+
+std::vector<Buffer> versionedKeys(const std::vector<std::string>& keys, const std::vector<BlockId>& versions) {
+  auto versioned_keys = std::vector<Buffer>{};
+  versioned_keys.reserve(keys.size());
+  std::transform(
+      keys.begin(), keys.end(), versions.begin(), std::back_inserter(versioned_keys), [](auto& key, auto version) {
+        auto key_hash = KeyHash{Hasher{}.digest(key.data(), key.size())};
+        return serialize(VersionedKey{key_hash, version});
+      });
+  return versioned_keys;
+}
+
+BlockMerkleCategory::BlockMerkleCategory(const std::shared_ptr<storage::rocksdb::NativeClient>& db) : db_{db} {
+  createColumnFamilyIfNotExisting(BLOCK_MERKLE_INTERNAL_NODES_CF, *db);
+  createColumnFamilyIfNotExisting(BLOCK_MERKLE_LEAF_NODES_CF, *db);
+  createColumnFamilyIfNotExisting(BLOCK_MERKLE_LATEST_KEY_VERSION, *db);
+  createColumnFamilyIfNotExisting(BLOCK_MERKLE_KEYS_CF, *db);
+  tree_ = sparse_merkle::Tree{std::make_shared<Reader>(*db_)};
+}
+
+BlockMerkleOutput BlockMerkleCategory::add(BlockId block_id, BlockMerkleInput&& updates, NativeWriteBatch& batch) {
+  auto merkle_value = hashUpdate(updates);
+  auto root_hash = merkle_value.root_hash;
+  putKeys(batch, block_id, merkle_value.hashed_added_keys, merkle_value.hashed_deleted_keys, updates);
+
+  auto block_key = serialize(BlockKey{block_id});
+  auto merkle_key = Sliver::copy((const char*)block_key.data(), block_key.size());
+  auto ser_value = serialize(merkle_value);
+  auto ser_value_sliver = Sliver::copy((const char*)ser_value.data(), ser_value.size());
+  auto tree_update_batch = tree_.update(SetOfKeyValuePairs{{merkle_key, ser_value_sliver}});
+
+  auto tree_version = tree_update_batch.stale.stale_since_version.value();
+  putMerkleNodes(batch, std::move(tree_update_batch), tree_version);
+
+  auto output = inputToOutput(updates);
+  output.root_hash = root_hash;
+  output.state_root_version = tree_version;
+  return output;
+}
+
+std::optional<MerkleValue> BlockMerkleCategory::get(const std::string& key, BlockId block_id) const {
+  auto hasher = Hasher{};
+  auto hashed_key = hasher.digest(key.data(), key.size());
+  return get(hashed_key, block_id);
+}
+
+std::optional<MerkleValue> BlockMerkleCategory::get(const Hash& hashed_key, BlockId block_id) const {
```

Comment:
Would that method return a value even if it has been deleted at `block_id`? I guess it is a tradeoff as you've pointed out, i.e. using a variant (with an overhead) to have a tombstone in the values table. Please let me know if I miss something.

Reply:
It cannot have been deleted **at** `block_id` if the key exists, but only earlier or later. Since the user is asking specifically for the value at `block_id` we return it if one exists.